tests/key_reuse_test.py

```diff
@@ -0,0 +1,301 @@
+      random.split(k)
+
+    with self.assertRaises(KeyReuseError):
+      jax.vmap(f)(keys)
+
+  def test_vmap_broadcast_key_double_use_errors(self):
+    key = random.PRNGKey(666)
+    def f(k, _):
+      random.split(k)
+      random.split(k)
+
+    with self.assertRaises(KeyReuseError):
+      jax.vmap(f, in_axes=(None, 0))(key, jnp.ones((2,)))
+
+  def test_scan_double_use_zero_iteration_errors(self):
+    # This case is rejected, even if this is a 0 iteration loop and the body
+    # technically does not execute,
+    def body(key, _):
+      random.split(key)
+      random.split(key)
+    key = random.PRNGKey(666)
+    with self.assertRaises(KeyReuseError):
+      jax.lax.scan(body, key, None, length=0)
+
+  def test_different_n_splits_errors(self):
+    key = random.PRNGKey(666)
+    random.split(key, num=2)
+    with self.assertRaises(KeyReuseError):
+      random.split(key, num=3)
+
+  def test_rand_bits_and_split_errors(self):
+    key = random.PRNGKey(666)
+    random.randint(key, (), 0, 10)
+    with self.assertRaises(KeyReuseError):
+      random.split(key, num=2)
+
+  # ---- does not error (correctly) ----
+  @parameterized.named_parameters(*RANDOM_FUNCS)
+  @jtu.ignore_warning(message="jax.random.shuffle is deprecated*")
+  def test_single_use_while_tracing(self, rand_function):
+    def single_use(key):
+      rand_function(key)
+      return 0.
+
+    key = random.PRNGKey(666)
+    jax.jit(single_use)(key)
+
+  def test_cond_uses_on_both_branches(self):
+    # This does not error because only one branch is taken.
+    def f(x, key):
+      return jax.lax.cond(x, random.split, random.split, key)
+    key = random.PRNGKey(777)
+    jax.jit(f)(True, key)
+
+  def test_consuming_slice_different_index(self):
+    # If a broadcast is a copy, this should not error.
+    # This can be thought of as a scan/map which maps over a broadcasted key
+    key = random.PRNGKey(666)
+    keys = jnp.broadcast_to(key, (2,))
+    random.split(keys[0])
+    random.split(keys[1])
+
+  def test_consuming_reference(self):
+    key = random.PRNGKey(666)
+    keys = jnp.broadcast_to(key, (2,))
+    random.split(key)
+    random.split(keys[0])
+
+  def test_scan_reuses_inp(self):
+    def body(carry, key):
+      random.split(key)
+      return carry, None
+    key = random.PRNGKey(666)
+    keys = jnp.broadcast_to(key, (2,))
+    jax.lax.scan(body, None, keys)
+
+  def test_vmap_broadcast_key_cross_batch(self):
+    key = random.PRNGKey(666)
+    def f(k, _):
+      random.split(k)  # Splitting same key across 2 batches.
+
+    # TODO(lenamartens): Is in_axes=(None,) an explicit broadcast?
+    # Should this behave similar to scan_reusing_inp?
+    jax.vmap(f, in_axes=(None, 0))(key, jnp.ones((2,)))
+
+  # TODO(lenamartens) fix false positives and false negatives.
+  # ---- false positives ----
+  @unittest.expectedFailure
+  def test_cond_uses_on_both_branches_closed(self):
+    # TODO(lenamartens): Is this a false positive? Yes, if we allow re-use on
```

Comment:
If it does not complicate the checking too much, I think that it is good to allow re-use on different branches of a conditional.